File: ensark-flutter/lib/models/account/account_models.dart
// ignore_for_file: invalid_annotation_target
import 'package:freezed_annotation/freezed_annotation.dart';
import '../enums.dart';

part 'account_models.freezed.dart';
part 'account_models.g.dart';

@freezed
class AccountRequest with _$AccountRequest {
  const factory AccountRequest({
    AccountType? accountType,
    double? availableBalance,
    int? branchId,
    @JsonKey(name: 'n_name') String? nName,
    @JsonKey(name: 'n_email') String? nEmail,
    @JsonKey(name: 'n_phone') String? nPhone,
    NomineeRelation? relation,
    @JsonKey(name: 'n_photo') String? nPhoto,
    @JsonKey(name: 'n_nid_front') String? nNidFront,
    @JsonKey(name: 'n_nid_back') String? nNidBack,
    List<AccountHolderRequest>? accountHolders,
  }) = _AccountRequest;

  factory AccountRequest.fromJson(Map<String, dynamic> json) => _$AccountRequestFromJson(json);
}

@freezed
class AccountResponse with _$AccountResponse {
  const factory AccountResponse({
    int? id,
    String? accountNumber,
    AccountType? accountType,
    AccountStatus? accountStatus,
    double? availableBalance,
    double? currentBalance,
    double? holdBalance,
    String? branchName,
    String? branchRoutingNumber,
    @JsonKey(name: 'n_name') String? nName,
    @JsonKey(name: 'n_email') String? nEmail,
    NomineeRelation? relation,
    @JsonKey(name: 'n_phone') String? nPhone,
    @JsonKey(name: 'n_photo') String? nPhoto,
    @JsonKey(name: 'n_nid_front') String? nNidFront,
    @JsonKey(name: 'n_nid_back') String? nNidBack,
    @Default([]) List<AccountHolderResponse> holderResponses,
  }) = _AccountResponse;

  factory AccountResponse.fromJson(Map<String, dynamic> json) => _$AccountResponseFromJson(json);
}

@freezed
class AccountHolderRequest with _$AccountHolderRequest {
  const factory AccountHolderRequest({
    HolderType? holderType,
    bool? canWithdraw,
    bool? canDeposit,
    bool? canApproveTransaction,
    String? signature,
    int? customerId,
  }) = _AccountHolderRequest;

  factory AccountHolderRequest.fromJson(Map<String, dynamic> json) => _$AccountHolderRequestFromJson(json);
}

@freezed
class AccountHolderResponse with _$AccountHolderResponse {
  const factory AccountHolderResponse({
    int? id,
    String? accountHolderName,
    HolderType? holderType,
    bool? canWithdraw,
    bool? canDeposit,
    String? signature,
    bool? canApproveTransaction,
  }) = _AccountHolderResponse;

  factory AccountHolderResponse.fromJson(Map<String, dynamic> json) => _$AccountHolderResponseFromJson(json);
}

File: ensark-frontend/src/app/core/models/account.models.ts
import { AccountType, AccountStatus, AccountCategory, NomineeRelation, HolderType } from './enums';

export interface AccountHolderRequest {
  holderType: HolderType;
  canWithdraw: boolean;
  canDeposit: boolean;
  canApproveTransaction: boolean;
  signature?: string;
  customerId: number;
}

export interface AccountRequest {
  accountType: AccountType;
  availableBalance: number;
  branchId: number;
  n_name: string;
  n_email: string;
  n_phone: string;
  relation: NomineeRelation;
  n_photo?: string;
  n_nid_front?: string;
  n_nid_back?: string;
  accountHolders: AccountHolderRequest[];
}

export interface AccountHolderResponse {
  id: number;
  accountHolderName: string;
  holderType: HolderType;
  canWithdraw: boolean;
  canDeposit: boolean;
  canApproveTransaction: boolean;
  signature?: string;
}

export interface AccountResponse {
  id: number;
  accountNumber: string;
  accountType: AccountType;
  accountStatus: AccountStatus;
  availableBalance: number;
  currentBalance: number;
  holdBalance: number;
  branchName: string;
  branchRoutingNumber: string;
  n_name?: string;
  n_email?: string;
  relation?: NomineeRelation;
  n_phone?: string;
  n_photo?: string;
  n_nid_front?: string;
  n_nid_back?: string;
  holderResponses: AccountHolderResponse[];
}
